ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/BigSamplerAvro.scala

```diff
@@ -38,67 +38,111 @@ import scala.concurrent.Future
                                                       hasher: Hasher): Hasher = {
-    val recordSchema = getRecordSchema(schema)
-    val subfields = fieldStr.split(BigSampler.fieldSep)
-    val field = recordSchema.getFields.asScala.find(_.name == subfields.head).getOrElse {
-      throw new NoSuchElementException(s"Can't find field $fieldStr in avro schema $schema")
-    }
-    val v = r.get(subfields.head)
+    hashAvroField(schema, r, fieldStr.split(BigSampler.fieldSep).toList, hasher)
+  }
+
+  @tailrec
+  private[samplers] def hashAvroField(schema: Schema, r: GenericRecord,
+                                                      fieldPath: List[String],
+                                                      hasher: Hasher): Hasher = {
+    val (field, v) = getField(r, fieldPath, schema)
     if (v == null) {
       log.debug(s"Field `${field.name}` of type ${field.schema.getType} is null - won't account" +
         s" for hash")
       hasher
     } else {
       field.schema.getType match {
         case Type.RECORD =>
-          hashAvroField(field.schema)(
+          hashAvroField(field.schema,
             v.asInstanceOf[GenericRecord],
-            subfields.tail.mkString(BigSampler.fieldSep.toString),
+            fieldPath.tail,
             hasher)
         case Type.UNION => hashAvroUnionField(field, v, hasher)
         case Type.ARRAY => hashAvroArrayField(field, v, hasher)
-        //  Type.MAP =>
+        case Type.MAP => hashAvroMapField(field, v, hasher)
         case t => hashPrimitive(field, t, v, hasher)
       }
     }
   }
   // scalastyle:on cyclomatic.complexity
-  private def hashBytes(bytes: Array[Byte], hasher: Hasher): Hasher = {
-    // Convert to string to have hex representations of bytes and raw bytes consistent
-    hasher.putString(Hex.encodeHexString(bytes), BigSampler.utf8Charset)
-  }
 
-  // scalastyle:off cyclomatic.complexity
   private def hashAvroArrayField(field: Schema.Field, v: AnyRef, hasher: Hasher): Hasher = {
-    field.schema.getElementType.getType match {
-      case Type.NULL =>
-        // ignore nulls
-        hasher
-      case t => v.asInstanceOf[JList[AnyRef]].asScala.foldLeft(hasher)((hasher, e) =>
-        hashPrimitive(field, t, e, hasher))
-    }
+    v.asInstanceOf[JList[AnyRef]].asScala.foldLeft(hasher)((hasher, e) =>
+        hashPrimitive(field, field.schema.getElementType.getType, e, hasher))
   }
 
-  // scalastyle:off cyclomatic.complexity
   private def hashAvroUnionField(field: Schema.Field, v: AnyRef, hasher: Hasher): Hasher = {
     val types = field.schema.getTypes.asScala
 
     types.foldLeft(hasher)( (hasher, p) =>
       p.getType match {
-        case Type.NULL =>
-          // ignore nulls
-          hasher
         case t => hashPrimitive(field, t, v, hasher)
       }
     )
   }
-  // scalastyle:on cyclomatic.complexity
+
+  private def hashAvroMapField(field: Schema.Field, v: AnyRef, hasher: Hasher): Hasher = {
```

Comment:
I wonder if we should have some kind of warning in the logs when hashing Map and Array fields. Hopefully mitigates amount of people shooting themselves in the foot with order issues.

Reply:
I agree and removed support for Map types, it seems too high risk that you would make a subtle order error. I also changed so that Ratatool handles repeated fields similarly in both Avro and BigQuery i.e. you can sample on them but not stratify on them. Array should be fine right, the order is guaranteed?